src/theory/sets/theory_sets_private.cpp

```diff
@@ -653,6 +667,116 @@ void TheorySetsPrivate::checkUpwardsClosure()
   }
 }
 
+void TheorySetsPrivate::checkMapUp()
+{
+  NodeManager* nm = NodeManager::currentNM();
+  const context::CDHashSet<Node>& mapTerms = d_state.getMapTerms();
+
+  for (const Node& term : mapTerms)
+  {
+    Node f = term[0];
+    Node A = term[1];
+    const std::map<Node, Node>& positiveMembers = d_state.getMembers(A);
+    shared_ptr<context::CDHashSet<Node>> skolemElements =
+        d_state.getMapSkolemElements(term);
+    for (const std::pair<const Node, Node>& pair : positiveMembers)
+    {
+      Node x = pair.first;
+      if (skolemElements->contains(x))
+      {
+        // Break this cycle between inferences SETS_MAP_DOWN_POSITIVE
+        // and SETS_MAP_UP:
+        // 1- If (set.member y (set.map f A)) holds, then SETS_MAP_DOWN_POSITIVE
+        //    inference would generate a fresh skolem x1 such that (= (f x1) y)
+        //    and (set.member x1 A).
+        // 2- Since (set.member x1 A) holds, SETS_MAP_UP would infer
+        //    (set.member (f x1) (set.map f A)).
+        // Since (set.member (f x1) (set.map f A)) holds, step 1 would repeat
+        // and generate a new skolem x2 such that (= (f x2) (f x1)) and
+        // (set.member x1 A). The cycle continues with step 2.
+        continue;
+      }
+      // (=
+      //   (and (set.member x B) (= A B))
+      //   (set.member (f x) (set.map f A))
+      // )
+      Node premise = pair.second;
+      Node B = pair.second[1];
+      if(A != B)
+      {
+        premise = premise.andNode(A.eqNode(B));
+      }
+      Node f_x = nm->mkNode(APPLY_UF, f, x);
+      Node skolem = d_treg.getProxy(term);
+      Node memberMap = nm->mkNode(kind::SET_MEMBER, f_x, skolem);
+      Node equivalence = premise.eqNode(memberMap);
+      std::vector<Node> empty;
+      d_im.assertInference(equivalence, InferenceId::SETS_MAP_UP, empty);
```

Comment:
Note that having a Boolean IFF as the conclusion necessarily makes the inference a lemma, not a fact.

It might be better to conclude the RHS equality with the first 2 equalities as an explanation (making this `=>` not `=`).  Can you do this instead?

Reply:
Done.